### PR TITLE
new approach to field publicizing in Inliner

### PR DESCRIPTION
**Pull request by magarciaEPFL**
Reworked version derived from

https://groups.google.com/d/topic/scala-internals/iPkMCygzws4/discussion

In particular,

https://groups.google.com/d/msg/scala-internals/iPkMCygzws4/LZHydgSjFSUJ
https://groups.google.com/d/msg/scala-internals/iPkMCygzws4/vEjJ4unAoFEJ
https://groups.google.com/d/msg/scala-internals/iPkMCygzws4/MIDPK6-3vwMJ

review by @VladUreche

Comments on performance impact are welcome /cc @gkossakowski

### CODE REVIEW

**Comment by gkossakowski** [comment]
Any chance for tests? We really need inliner tests and we have infrastructure to write them now (see 4caa766e85b).

File: src/compiler/scala/tools/nsc/backend/opt/Inliners.scala [hunk 1]
```diff
@@ -645,31 +645,40 @@ abstract class Inliners extends SubComponent {
...
-         
-        def potentiallyPublicized(f: Symbol): Boolean = {
-          (m.sourceFile eq NoSourceFile) && f.name.containsChar('$')
+         **/
+        def canRegardAsPublic(f: Symbol): Boolean = {
+
+          val calleeSym = m.symbol
+          if(hasInline(calleeSym)) {
+            if(currentRun.compiles(f)) {
+              toBecomePublic ::= f;
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/backend/opt/Inliners.scala:663]: This should already be public after superaccessors, no? If so, why add them to toBecomePublic? I'd rather do `assert(f.isPublic /* or whatever the incantation is*/)`.

File: src/compiler/scala/tools/nsc/backend/opt/Inliners.scala [hunk 2]
```diff
@@ -645,31 +645,40 @@ abstract class Inliners extends SubComponent {
...
+          }
+          val canForcePublic = {
+            currentRun.compiles(f) && (
+              f.isSynthetic     ||
+              f.isParamAccessor ||
+              (isClosureClass(f.owner) && f.isOuterField /* ie is an anon-closure-class field */ )
+            )
+          }
+          if(canForcePublic) {
+            toBecomePublic ::= f;
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/backend/opt/Inliners.scala:675]: This public status is only valid in the current compilation run, right? We need to check that the pickled signature says it's private, else the next compilation run will link against members which are not always public, depending on how you compile.

**Comment by VladUreche** [comment]
And as Greg pointed out, I'd love to see some tests for it.

**Comment by magarciaEPFL** [comment]
As I said I'm focusing first on finding out what's wrong with the approach, tests afterwards.

Looking at the conditions under which `canRegardAsPublic(f)` is invoked:

```
def checkField(f: Symbol)   = check(f, f.isPrivate && !canRegardAsPublic(f))
```

The following assertion should hold on entry to `canRegardAsPublic(f)`

```
assert(!hasInline(calleeSym),
       "All accesses in @inline-marked methods should have resulted in ExplicitOuter publicizing the accessed field.")
```

I'll check whether it actually holds. Probably not, because on second thought, I guess #1121 should have been stated as follows:

```
      if (currentClass != sym.owner ||
          (currentMethod hasAnnotation ScalaInlineClass)) {
        assert(currentRun.compiles(sym), "Trying to publicize a separately-compiled field.")
        sym.makeNotPrivate(sym.owner)  
      }
```

WDYT?

**Comment by VladUreche** [comment]
You're right, it probably doesn't hold, but for a completely different reason -- the entry might be protected instead of private. I think the check should be `!f.isPublic`.

Regarding the patch in #1121, you're right, it should check it's in the same compilation unit, else it will over-approximate the fields that are public.

**Comment by magarciaEPFL** [comment]
The following reformulation of #1121 seems all we can do in ExplicitOuter:

```
      // make not private symbol acessed from inner classes, as well as
      // symbols accessed from @inline methods
      if ((currentMethod hasAnnotation ScalaInlineClass) && currentRun.compiles(sym)) {
        sym.makeNotPrivate(sym.owner)
        if(sym.isProtected) { sym.setFlag(notPROTECTED) }
      }
      else if (currentClass != sym.owner) {
        sym.makeNotPrivate(sym.owner)
      }
```

Better yet, the idea of `closestEnclMethod()` from scala/scala@2fd4df89ce12ed84852f11effda1d42bf42acc54 could be brought back, because in case `currentMethod` isn't a `sourceMethod`, it walks up till one is found.

**Comment by VladUreche** [comment]
I didn't follow the member public-izing problem closely, so I don't exactly know all the aspects. Why do you do:

```
      else if (currentClass != sym.owner) {
        sym.makeNotPrivate(sym.owner)
      }
```

?

**Comment by magarciaEPFL** [comment]
That pre-existed publicizing-for-inlining, it's there so that after flatten has run, accesses from an inner-class reaching out to outer members also work once the inner-class isn't an inner class anymore (after flattten).

**Comment by magarciaEPFL** [comment]
WIP: The above seems to work, yet doesn't quite do what we want: there are outer fields accessed in at-inline methdos that reach Inliner being private.

**Comment by VladUreche** [comment]
Thanks for the clarification Miguel.

**Pull request closed without merging**